### PR TITLE
fix(cli): don't walk the subdirectory twice when using the `--ignore` flag

### LINKED ISSUE #7736: deno lint/fmt --ignore options is very slow
**Overview**
Was confused why `deno lint` and `deno fmt` took 50seconds to finish 

**What Is The Bug**
`deno lint` and `deno fmt` still check the ignored files (`--ignore=...`) which isn't as expected, but do not print to stdout, which is expected

I know this by following the steps below

**How to Reproduce**
Here is my **exact example**, feel free to change it around if you wish. The focal point being having a directory with lots and lots of files and lots and lots of code

1) Make an empty directory and cd into it

2) Run `time deno lint --unstable` and `time deno fmt`. **Make note of the times**

3) Create a `package.json` with bunch of dependencies:

```
{
  "name": "whatever",
  "author": "whatever",
  "version": "0.0.1",
  "devDependencies": {
    "@babel/core": "^7.3.4",
    "acorn": ">=5.7.4",
    "axios": "^0.19.2",
    "babel-cli": "^6.26.0",
    "babel-loader": "^8.0.5",
    "babel-preset-env": "^1.7.0",
    "braces": ">=2.3.1",
    "css-loader": "^2.1.0",
    "date-fns": "^2.14.0",
    "kind-of": ">=6.0.3",
    "markdown-it": "^8.4.2",
    "mime-types": "^2.1.22",
    "node-sass": "^4.13.1",
    "prismjs": "^1.15.0",
    "pug": "^2.0.3",
    "pug-plain-loader": "^1.0.0",
    "sass-loader": "^7.1.0",
    "scss-loader": "0.0.1",
    "style-loader": "^0.23.1",
    "sweetalert": "^2.1.2",
    "typedoc": "^0.15.6",
    "vue": "^2.6.7",
    "vue-axios": "^2.1.5",
    "vue-loader": "^15.6.4",
    "vue-router": "^3.0.2",
    "vue-style-loader": "^4.1.2",
    "vue-template-compiler": "^2.6.7",
    "vuex": "^3.4.0",
    "webpack": "^4.29.5",
    "webpack-cli": "^3.2.3"
  },
  "scripts": {
    "webpack": "node_modules/.bin/webpack --hide-modules --env.environment=development",
    "webpack-watch": "node_modules/.bin/webpack --hide-modules --watch --env.environment=development"
  },
  "dependencies": {
    "@johmun/vue-tags-input": "^2.1.0"
  }
}
```

4) Run `npm i` to install the dependencies

5) Run `time deno lint --unstable --ignore=node_modules` and `time deno fmt --ignore=node_modules`. Make note of the times

6) Compare times from 2nd and. 5th step

7) Notice the second result is **significantly longer**

And to truly know these commands still check the files,  just `rm -rf. node_modules`, and run the `time <command>` again

### PR DESCRIPTION
**Pull request by humancalico**
Fixes #7736

### CODE REVIEW

**Comment by bartlomieju** [comment]
@nayeemrmn could you review it?

File: cli/fmt.rs [hunk 1]
```diff
@@ -234,21 +227,41 @@ fn is_supported(path: &Path) -> bool {
...
 
   if files.is_empty() {
-    target_files.extend(files_in_subtree(
-      canonicalize_path(&std::env::current_dir()?)?,
-      is_supported,
-    ));
+    for entry in WalkDir::new(std::env::current_dir()?.canonicalize()?)
+      .into_iter()
+      .filter_entry(|e| {
+        !ignore.contains(&e.clone().path().canonicalize().unwrap())
```
Comment (nayeemrmn) [inline cli/fmt.rs:243]: This isn't sufficient to match the current ignore semantics. Path prefixes are supported.

File: cli/fmt.rs [hunk 2]
```diff
@@ -234,21 +227,41 @@ fn is_supported(path: &Path) -> bool {
 
 pub fn collect_files(
   files: Vec<PathBuf>,
+  ignore: Vec<PathBuf>,
 ) -> Result<Vec<PathBuf>, std::io::Error> {
   let mut target_files: Vec<PathBuf> = vec![];
+  // FIXME the unwrap here panicks when we add a ignore file/folder which doesn't exist
```
Comment (nayeemrmn) [inline cli/fmt.rs:233]: Regression? Should be addressed here then.

**Review by nayeemrmn — CHANGES_REQUESTED** [review]
On filtering, you have to check if each entry `starts_with` any ignored path.

File: cli/fmt.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/fmt.rs:233]: Non-existent paths passed to `ignore` should be skipped

**Comment by bartlomieju** [comment]
@nayeemrmn please give it another pass

**Review by nayeemrmn — APPROVED** [review]
LGTM, if the performance is satisfactory @bartlomieju.

**Review by bartlomieju — APPROVED** [review]
I verified that performance is significantly better, thanks @humancalico great fix!

**Pull request merged**